typing/typecore.ml

```diff
@@ -4542,13 +4542,18 @@ and type_cases ?in_function env ty_arg ty_res partial_flag loc caselist =
   (* ty_arg is _fully_ generalized *)
   let patterns = List.map (fun {pc_lhs=p} -> p) caselist in
   let contains_polyvars = List.exists contains_polymorphic_variant patterns in
-  let erase_either = contains_polyvars && contains_variant_either ty_arg
-  and has_constructors = List.exists contains_constructor patterns in
+  let erase_either = contains_polyvars && contains_variant_either ty_arg in
+  (* There are various things that we need to do in presence of GADT
+     constructors that aren't required if there are none.
+     However, because of disambiguation, we can't know for sure whether the
+     patterns contain some GADT constructors. So we conservatively assume that
+     any constructor might be a GADT constructor. *)
+  let may_contain_gadts = List.exists contains_constructor patterns in
```

Comment:
I like this patch but I think that you may go even further and rename `contains_constructor` into `may_contain_gadts`, and call the local boolean `has_gadts` again (or `may_have_gadts` if you wish). This would make other call sites, such as the Pexp_let one, easier to follow.

Reply:
I guess I could rename the function and move the comment to its definition. I don't like `has_gadts` but `may_have_gadts` would be fine if you prefer that over `contain`.

Reply:
I have no opinion over `have/has` vs `contain/contains`, just that the toplevel function name could express that it's actually used to look for potential gadts. No opinion on what the local variable name should be (as long as it doesn't shadow the function name, that's ugly) -- reusing `has_gadts` had the advantage of a smaller diff but that does not matter so much.